### ISSUE TITLE
Rust-analyzer inserts & (ampersand) character on the wrong line when auto-completing variable reference

### ISSUE THREAD

**Issue by stevepryde** [rust-lang/rust-analyzer#12717]
When I auto-complete a function parameter that wants a reference, I get a `&` character randomly inserted on the wrong line in the same file.

It appears to use the line-number relative to the file rather than relative to the current function.

See below for code and steps to reproduce the issue.

**rust-analyzer version**: (eg. output of "Rust Analyzer: Show RA Version" command)
rust-analyzer version: 0.0.0 (75b22326d 2022-07-03)

**rustc version**: (eg. output of `rustc -V`)
rustc 1.62.0 (a8314ef7d 2022-06-27)

```rust
#[derive(Debug)]
pub struct QueryResultRequest {
    query_id: String,
}

pub async fn query_result(request: QueryResultRequest) -> Result<String, String> {
    Ok(request.query_id)
}

pub fn show_value(value: &str) {
    println!("value: {value}");
}

fn main() {
    println!("Hello, world!");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test() {
        let request = QueryResultRequest {
            query_id: "test".to_string(),
        };

        query_result(request).await.unwrap();
    }
}
```

Steps to reproduce:
1. Locate the line of code at the bottom (`query_result(...)`)
2. Above that line, type the following, but auto-complete the variable name
`show_value(request.que` (then press ENTER to auto-complete `query_id`)

Upon auto-completion, observe an `&` character has been inserted on line 6 of the file, making the function signature now look like this:
```rust
pub async fn query_result(request: QueryResultRe&quest) -> Result<String, String> {
                                                ^
```

**Comment by stevepryde** [comment]
if I change `#[tokio::test]` to `#[test]` and remove the async etc. I can't reproduce the issue. May be related to that macro, not sure.

**Comment by Veykril** [comment]
Ah we don't consider macros for these ye, these are all the problematic locations
https://github.com/rust-lang/rust-analyzer/blob/09abd76c6a601797c2287a9ef341b7ce904e8aa6/crates/ide-completion/src/render.rs#L138
https://github.com/rust-lang/rust-analyzer/blob/09abd76c6a601797c2287a9ef341b7ce904e8aa6/crates/ide-completion/src/render.rs#L317
https://github.com/rust-lang/rust-analyzer/blob/09abd76c6a601797c2287a9ef341b7ce904e8aa6/crates/ide-completion/src/render/function.rs#L82-L92
https://github.com/rust-lang/rust-analyzer/blob/09abd76c6a601797c2287a9ef341b7ce904e8aa6/crates/ide-completion/src/render/literal.rs#L124-L126

We'll need to use https://github.com/rust-lang/rust-analyzer/blob/09abd76c6a601797c2287a9ef341b7ce904e8aa6/crates/hir/src/semantics.rs#L250-L254 for this I believe, if we get None back we shouldn't emit this special completion for now.

**Comment by abusch** [comment]
I can also reproduce this by typing `show_value(request.query_id);` (without using completion), then executing the "Add reference here" code action, so it doesn't seem to only concern completion.

**Comment by Veykril** [comment]
thats the same issue in a different part of the project, feel free to create a new issue regarding that

**Comment by lnicola** [comment]
I think this was fixed in #12800 and #12830.
